src/storage/garbage_collector.cpp

```diff
@@ -87,32 +122,111 @@ uint32_t GarbageCollector::ProcessUnlinkQueue() {
+        requeue.push_front(txn);
+      }
+    }
+  }
+
+  storage::UndoBuffer *buf = nullptr;
+  std::forward_list<storage::UndoBuffer *> buf_requeue;
+
+  // Process every compaction buffer in the buffer unlink queue
+  while (!buffers_to_unlink_.empty()) {
+    buf = buffers_to_unlink_.front();
+    buffers_to_unlink_.pop_front();
+    if (buf->Empty()) {
+      // This compaction buffer is empty so it is safe to delete it
+      delete buf;
     } else {
-      // This is a committed txn that is still visible, requeue for next GC run
-      requeue.push_front(txn);
+      // This is a list of undo records that may or may not be visible to any running txns. Proceed with unlinking them
+      // with an Interval GC approach
+      bool all_unlinked = true;
+      for (auto &undo_record : *buf) {
+        all_unlinked = ProcessUndoRecord(&undo_record, &active_txns, &visited_slots) && all_unlinked;
+      }
+      if (all_unlinked) {
+        // We unlinked all of the UndoRecords for this compaction buffer, so we can add it to the deallocation queue
+        buffers_to_deallocate_.push_front(buf);
+      } else {
+        // We didn't unlink all of the UndoRecords (UnlinkUndoRecord returned false due to a write-write conflict),
+        // requeue buf for next GC run. Unlinked UndoRecords will be skipped on the next time around since we use the
+        // table pointer of an UndoRecord as the internal marker of being unlinked or not
+        buf_requeue.push_front(buf);
+      }
     }
   }
 
-  // Requeue any txns that we were still visible to running transactions
-  txns_to_unlink_ = transaction::TransactionQueue(std::move(requeue));
+  // Requeue any txns that have an undo record still visible to some running transaction
+  if (!requeue.empty()) {
+    txns_to_unlink_ = transaction::TransactionQueue(std::move(requeue));
+  }
+
+  // Requeue any compaction buffers that that have an undo record still visible to some running transaction
+  if (!buf_requeue.empty()) {
+    buffers_to_unlink_ = std::forward_list<storage::UndoBuffer *>(std::move(buf_requeue));
+  }
 
   return txns_processed;
 }
 
+bool GarbageCollector::ProcessUndoRecord(UndoRecord *const undo_record,
+                                         std::vector<transaction::timestamp_t> *const active_txns,
+                                         std::unordered_set<TupleSlot> *visited_slots) {
+  DataTable *table = undo_record->Table();
+  // If this UndoRecord has already been processed, we can skip it
+  if (undo_record->txnptr_.IsNull()) return true;
+  const TupleSlot slot = undo_record->Slot();
+  // Process this tuple only
+  if (visited_slots->insert(slot).second) {
+    // Perform interval gc for the entire version chain excluding the head of the chain
+    ProcessTupleVersionChain(table, slot, active_txns);
+    ProcessTupleVersionChainHead(table, slot, active_txns);
+  }
+
+  return undo_record->txnptr_.IsNull();
```

Comment:
Question: Is this line just going to be false, or `ProcessTupleVersionChain` and `ProcessTupleVersionChainHead` can modified undo_record to change that fact? If it is the former, might as well just return false.

Reply:
undo records have an extra field containing the pointer to the corresponding txn. This txn pointer is set to NULL when the undo record is unlinked. So this line is not necessarily false.